File: phylotyper/R/phylotyper.R
##########################################################
## File: phylotyper.R
##
##  Predict subtype state for a small number of unknowns in
##  the subtype DNA sequence tree
##
## 
## 
##
##
##
## author: Matt Whiteside
## copyright: Copyright Government of Canada 2012-2015. Funded by the Government of Canada Genomics Research and Development Initiative
## license: ASL
## version: 2.0
## maintainer: Matt Whiteside
## email: mdwhitesi@gmail.com
##
##########################################################

wd <- setwd(".")
setwd(wd)

# Create enviroment to keep phylotyper functions
phylotyper = new.env()


phylotyper$runSubtypeProcedure <- function(tree, priorM, scheme=1, ...) {
	# A Wrapper that runs one of several possible subtype 
	# estimation procedures
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  priorM: a matrix of prior values indicating tip node subtype state
	#  scheme: an integer between 1:? indicating procedure to use.
	#    Possible procedures (phytool method, model):
	#      1=rerootingMethod, ER
	#      2=rerootingMethod, SYM
	#      3=rerootingMethod, ARD
	#      4=make.simmap, ER
	#      5=tip.posterior.probability, ER
	#
	# Returns list with:
	#   ?
	#

	if(scheme == 1) {
		# Rerooting method, ER model
		capture.output(fit <- rerootingMethod(tree, priorM, model='ER', tips=TRUE))
		return(list(result=fit, tip.pp=fit$marginal.anc[tree$tip.label,], plot.function='plotRR'))

	} else if(scheme == 2) {
		# Rerooting method, SYM model
		fit = rerootingMethod(tree, priorM, model='SYM', tips=TRUE)
		return(list(result=fit, tip.pp=fit$marginal.anc[tree$tip.label,], plot.function='plotRR'))

	} else if(scheme == 3) {
		# Rerooting method, ARD model
		fit = rerootingMethod(tree, priorM, model='ARD', tips=TRUE)
		return(list(result=fit, tip.pp=fit$marginal.anc[tree$tip.label,], plot.function='plotRR'))

	} else if(scheme == 4) {
		# simmap method, ER model
		trees = make.simmap(tree, priorM, model='ER', nsim=1000)
		fit = describe.simmap(trees, plot=FALSE)
		return(list(result=fit, tip.pp=fit$tips, plot.function='plotSM'))

	} else if(scheme == 5) {
		# tip.posterior.probability method, ER model or specified using model= parameter
		if(hasArg(tips)) tips<-list(...)$tips
		else stop('Missing argument: tips')

		model='ER'
		if(hasArg(model)) model<-list(...)$model

		fixedQ=NULL
		if(hasArg(fixedQ)) fixedQ<-list(...)$fixedQ
		
		fit = tip.posterior.probability(tree, priorM, tips, model=model, fixedQ=fixedQ)
		return(list(result=fit, tip.pp=fit$marginal.anc, plot.function='plotTPP'))

	}

}

phylotyper$makePriors <- function(tree, subtypes) {
	# Generate prior matrix used as input for ACE
	# 
	# Tree tip nodes with no subtype in x, will be assigned
	# flat priors (equal likelihood of having any subtype)
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  subtypes: factor list of subtype assignments. List names
	#    must match tip names in tree, except for tips
	#    with unknown subtypes (which subtypes
	#    will be predicted for)
	#
	# Returns list with:
	#   prior.matrix: matrix containing prior values
	#   untyped: vector of tree tip names with no subtype 
	#
	
	# Find leaves with no subtype
	undefined = setdiff(tree$tip.label, names(subtypes))

	# Build prior matrix
	priors <- to.matrix(subtypes,seq=levels(subtypes))

	# Add flat priors for leaves with no subtype
	numstates = length(levels(subtypes))
	if(length(undefined) > 0)
		priors <- rbind(priors, matrix(1/numstates,
			length(undefined), numstates, dimnames=list(undefined)))

	# Reorder based on tip order
	priors <- priors[tree$tip.label,]

	return(list(prior.matrix=priors, untyped=undefined))
}

phylotyper$loadInstallLibraries <- function(libloc="~/R/", repo="http://cran.stat.sfu.ca/") {
	# Attempt to load required libraries
	# If library not installed, attempt install
	#
	# Args:
	#  libloc: If defined, add path string to R's default internal Library paths variable
	#  repo: R repository to use in install.packages
	#
	# Returns:
	#   nothing
	#
	# Function will throw error if load/install fails
	# 

	# Location to install libraries to
	if(! is.null(libloc)) {
		.libPaths(c(libloc,.libPaths()))
	}
	
	# Set repo
	r = getOption("repos")
	r["CRAN"] = repo
	options(repos = r)

	# Install libraries from Bioconductor
	bioc.libs = c("Biostrings")
	if(!exists('biocLite')) {
		source("http://bioconductor.org/biocLite.R")
	}
	for(x in bioc.libs) {
		if (!require(x,character.only = TRUE)) {
	  		biocLite(x)
	    	if(!require(x,character.only = TRUE)) stop(paste("Package not loaded: ",x))
		}
	}
	
	# Install libraries from CRAN
	cran.libs = c("devtools", "ape", "phangorn", "RColorBrewer", "ggplot2", "optparse", "mclust",
		"robustbase", "fitdistrplus", "igraph")
	for(x in cran.libs) {
		if (!require(x,character.only = TRUE)) {
	  		install.packages(x,dep=TRUE)
	  		library(x)
	    	if(!require(x,character.only = TRUE)) stop(paste("Package not loaded: ",x))
		}
	}
	
	# Install libraries from Github
	git.libs = data.frame(lib=c("phytools"), git.path=c("liamrevell/phytools"))
	for(i in 1:nrow(git.libs)) {
    	lib = git.libs[i,1]
   		path = git.libs[i,2]
    	if (!require(lib, character.only = TRUE)) {
  			install_github(path)
    		if(!require(lib, character.only = TRUE)) stop(paste("Package not loaded:",lib))
		}
	}

	# Install older version of root, unroot, and is.root functions from APE packages
	# New version have bug
	source('root.R')
	# Reload methods that rely on root.R methods
	source('rerootingMethod.R')
	source('utilities.R')
}

phylotyper$loadSubtype <- function(treefile, stfile=NULL, do.root=TRUE, resolve.polytomies=TRUE) {
	# Load tree and subtype assignments associated with subtype scheme
	#
	# Root tree at midpoint.
	#
	# Args:
	#  treefile: Path to tree in newick format
	#  stfile: Path to subtype assignments in format: tree_tip_label\tsubtype
	#  do.root: Boolean indicating if midpoint.root should be called on tree
	#  resolve.polytomies: Boolean indicating if multifurcating nodes should be expanded to bifurcating
	#
	# Returns:
	#   list:
	#     tree = phylo object
    #     untyped = list of tree_tip_labels not found in stfile
    #     subtypes = factor of subtype assignments. Names() match tree_tip_labels
	#
	# 
	# 

	otree = read.tree(treefile)
	tree = otree
	if(do.root) tree <- midpoint.root(tree)
	if(resolve.polytomies) {
		tree <- multi2di(tree,random=TRUE)
	}

	res=list(tree=tree, multif=otree)

	if(!is.null(stfile)) {
		# Load subtypes
		st = read.table(stfile, sep="\t", row.names=1)

		# Convert to factor list
		subtypes = st[,1]
		names(subtypes) = rownames(st)
		res[['subtypes']] = subtypes

		# Make note of labels to find subtypes for
		undefined = setdiff(tree$tip.label, names(subtypes))
		res[['untyped']] = undefined
	}
	
	return(res)
}

phylotyper$mypalette <- function(subtypes) {
	# Generate a set of colors representing subtypes
	#
	# Args:
	#   subtypes: factor list of subtype assignments. List names
	#    must match tip names in tree, except for tips
	#    with unknown subtypes (which subtypes
	#    will be predicted for)
	#
	# Returns:
	#   list of color assignments. List names match subtype names
	# 
	# 

	
	states = levels(subtypes)
	n = length(states)
	cols = rainbow(n, s=0.6,v=1)

	if(n > length(cols)) {
		stop("Number of subtypes exceeds available colors in palette. Please defined your own color palette.")
	}

	pal = cols[sample(1:n,n)]
	names(pal) = states

	pal = c(pal,'other'="#D3D3D3")

	return(pal)
}

phylotyper$piecolors <- function(marginals) {

	n = ncol(marginals)
	cutoff = 1/n
	reduced = cbind(marginals,other=0)

	reduced[,'other'] = apply(marginals, 1, function(x) { sum(x[which(x <= cutoff)]) })
	reduced[reduced <= cutoff] = 0

	return(reduced)
}

phylotyper$plot.subtype <- function(tree, subtypes, tip.subtypes=NULL) {
	# plot tree with subtype assignments overlayed
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  subtypes: factor list of subtype assignments
	#
	# Returns:
	#   nothing
	#

	cols = cols2 = phylotyper$mypalette(subtypes)
	subtypes2 = subtypes
	diff.tips = FALSE
	if(!is.null(tip.subtypes)){
		subtypes2 = tip.subtypes
		cols2 = phylotyper$mypalette(subtypes2)
		diff.tips = TRUE
	}
	pies = matrix(0, ncol=length(cols),nrow=length(subtypes))
	colnames(pies) = names(cols)
	rownames(pies) = tree$tip.label
	for(i in 1:nrow(pies)) {
		st = subtypes[rownames(pies)[i]]
		pies[i,st] = 1
	}

	plot(tree,label.offset=0.001,cex=0.7,type='fan',align.tip.label=TRUE,tip.col=cols2[subtypes2[tree$tip.label]])

	tiplabels(pie=pies, 
		piecol=cols,
		cex=0.2)
	
	add.simmap.legend(colors=cols,x=0.9*par()$usr[2],
		y=0.9*par()$usr[4],prompt=FALSE)

	# if(diff.tips) {
	# 	add.simmap.legend(colors=cols2,x=0.7*par()$usr[2],
	# 	y=0.9*par()$usr[4],prompt=FALSE)
	# }
}

phylotyper$plot.subtype2 <- function(tree, subtypes, tip.subtypes=NULL) {
	# plot tree with subtype assignments overlayed
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  subtypes: factor list of subtype assignments
	#
	# Returns:
	#   nothing
	#

	cols = cols2 = phylotyper$mypalette(subtypes)
	subtypes2 = subtypes
	diff.tips = FALSE
	if(!is.null(tip.subtypes)){
		subtypes2 = tip.subtypes
		cols2 = phylotyper$mypalette(subtypes2)
		diff.tips = TRUE
	}
	pies = matrix(0, ncol=length(cols),nrow=length(subtypes))
	colnames(pies) = names(cols)
	rownames(pies) = tree$tip.label
	for(i in 1:nrow(pies)) {
		st = subtypes[rownames(pies)[i]]
		pies[i,st] = 1
	}

	plot(tree,label.offset=0.001,cex=0.7,type='phylogram',align.tip.label=TRUE,tip.col=cols2[subtypes2[tree$tip.label]])

	tiplabels(pie=pies, 
		piecol=cols,
		cex=0.2)

	nodelabels(cex=0.2)
	
	add.simmap.legend(colors=cols,x=0.9*par()$usr[2],
		y=0.9*par()$usr[4],prompt=FALSE)

	# if(diff.tips) {
	# 	add.simmap.legend(colors=cols2,x=0.7*par()$usr[2],
	# 	y=0.9*par()$usr[4],prompt=FALSE)
	# }
}

phylotyper$plot.anc <- function(tree, fit, subtypes) {
	# plot tree with internal node marginal anc overlayed
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  subtypes: factor list of subtype assignments
	#
	# Returns:
	#   nothing
	#

	cols = phylotyper$mypalette(subtypes)

	plot(tree,label.offset=0.001,cex=0.8,type='fan',align.tip.label=TRUE,tip.col=cols[subtypes[tree$tip.label]],
		no.margin=TRUE)

	nodelabels(pie=phylotyper$piecolors(fit$marginal.anc),
		piecol=cols,
		cex=0.6)
	
	add.simmap.legend(colors=cols,x=0.01*par()$usr[2],
		y=0.95*par()$usr[4],prompt=FALSE,cex=0.8)
} 

phylotyper$plotRR <- function(tree, fit, subtypes) {
	# plot tree with posterior probabilities from 
	# rerootingMethod function displayed as pies
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  fit: output object from rerootingMethod()
	#  subtypes: factor list of subtype assignments
	#
	# Returns:
	#   nothing
	#

	cols = phylotyper$mypalette(subtypes)

	plot(tree,label.offset=0.001,cex=0.7,type='fan',align.tip.label=TRUE,tip.col=cols[subtypes[tree$tip.label]])

	tiplabels(pie=fit$marginal.anc[tree$tip.label,], 
		piecol=cols,
		cex=0.2)

	nodelabels(pie=phylotyper$piecolors(fit$marginal.anc[as.character(1:tree$Nnode+Ntip(tree)),]),
		piecol=cols,
		cex=0.4)
	
	add.simmap.legend(colors=cols,x=0.9*par()$usr[2],
		y=0.9*par()$usr[4],prompt=FALSE)
}

phylotyper$plotSM <- function(tree, fit, subtypes) {
	# plot tree with posterior probabilities from 
	# make.simmap function displayed as pies
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  fit: output object from rerootingMethod()
	#  subtypes: factor list of subtype assignments
	#  plot.nodes: display interior tree node ancestral state
	#    posterior probability
	#
	# Returns:
	#   nothing
	#

	cols = phylotyper$mypalette(subtypes)

	plot(fit, colors=cols, fsize=0.6)

	add.simmap.legend(colors=cols,x=0.9*par()$usr[2],
		y=0.9*par()$usr[4],prompt=FALSE)
}

phylotyper$plotTPP <- function(fit, tree, subtypes) {
	# plot tree with tip posterior probability and conditional
	# likelihoods from tip.posterior.probability function
	#
	# Args:
	#  fit: output object from tip.posterior.probabilities()
	#  tree: not used, use fit$tree provided by the fit object
	#  subtypes: factor list of subtype assignments
	#
	# Returns:
	#   nothing
	#

	tree = fit$rerootedTree

	# Save original node assignments in the node.label field
	tree$node.label = as.character(1:tree$Nnode+Ntip(tree))

	# Reroot for better viewing
	tree = midpoint.root(tree)

	cols = phylotyper$mypalette(subtypes)
	probs = phylotyper$makePriors(tree, subtypes)$prior.matrix
	probs[rownames(fit$marginal.anc),] = fit$marginal.anc

	plot(tree,label.offset=0.001,cex=0.7,type='phylo',align.tip.label=TRUE,
		tip.col=cols[subtypes[tree$tip.label]], main=paste('Posterior probability and associated subtree conditional\nlikelihoods for tip',fit$tip))

	tiplabels(pie=probs[tree$tip.label,],
		piecol=cols,
		cex=0.2)

	original.nodes = tree$node.label[tree$node.label != "Root"]
	nstates = ncol(fit$conditional.likelihoods)
	root.prior = rep(1/nstates, nstates)
	probs2 = rbind(root.prior, fit$conditional.likelihoods[original.nodes,])
	nodelabels(pie=phylotyper$piecolors(probs2),
		piecol=cols,
		cex=0.4)
	
	add.simmap.legend(colors=cols,x=0.1*par()$usr[2],
		y=0.9*par()$usr[4],prompt=FALSE)

}


phylotyper$plotDim <- function(tree, type='fan') {
	# Compute X, Y dimensions for a plot to fit tree
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#
	# Returns:
	#   list with x, y, res names
	#

	x_scale = 3600
	y_scale = 16
	x_max = 1728
	y_max = 1728
	res = 72

	# X
	h = max(nodeHeights(tree))
	x = h * x_scale
	x = ifelse(x > x_max, x_max, x)

	if(type != 'fan') {
		# Y
		l = length(tree$tip.label)
		y = l * y_scale
		y = ifelse(y > y_max, y_max, y)
	} else {
		y <- x
	}

	return(list(x=x,y=y,res=res))
}

phylotyper$tip.posterior.probability <- function(tree,priorM,uncertain,model=c("ER","SYM"),fixedQ=NULL) {
	# Adapted from phytools rerootingMethod
	#
	# It computes the marginal posterior probabiliyt of a tip in tree by rooting at lowest point
	# on edge descending to tip. Unlike the original rerootingMethod that this function
	# is adapted from, it only iterates for each tip listed in the uncertain argument.
	#
	# Marginal ancestral state posterior probability is equivalent to the conditional 
	# scaled likelihoods at the root node of the tree (obtained by fitMk). The original 
	# rerootingMethod roots the tree at each internal node and tip edge to generate the 
	# marginals for each node tip in the tree.
	#
	# Args:
	#  tree: phylo object containing subtype tree
	#  priorM: a matrix of prior values indicating tip node subtype state
	#
	# Returns:
	#   list with x, y, res names
	#

	if(!inherits(tree,"phylo")) 
		stop("tree should be an object of class \"phylo\".")
	if(!is.matrix(model)) model<-model[1]
	n<-Ntip(tree)
	
	if(!is.matrix(priorM)){ 
		stop("priorM should be a matrix")
	}
	yy<-priorM
	yy<-yy[tree$tip.label,]
	yy<-yy/rowSums(yy)

	# Check the targets
	if(length(uncertain) < 1 || length(uncertain) > n) {
		stop('Missing / invalid argument: uncertain')
	}
	tl <- uncertain %in% tree$tip.label
	if(any(!tl)) {
		stop("Unrecognized tip label in uncertain argument")
	}
	nn = which(tree$tip.label %in% uncertain)
	if(any(nn > n)) {
		stop("Not a valid tip node in tree")
	}

	# Start with first target tip
	nn1 = nn[1]
	nn = nn[-1]

	# Compute marginal probability for first unknown tip and also transition matrix

	# Root tree at lowest point on edge to tip
	tt <- reroot(tree,nn1,tree$edge.length[which(tree$edge[,2]==nn1)])

	if(is.null(fixedQ)) {
		YY<-fitMk(tt,yy,model=model,output.liks=TRUE)
		Q <- phylotyper$makeQ(YY)
	} else {
		if(!is.matrix(fixedQ)){ 
			stop("fixedQ should be a matrix")
		}
		Q <- fixedQ
		YY<-fitMk(tt,yy,model=model,fixedQ=Q,output.liks=TRUE)
	}
	
	# Repeat for remaining tips
	ff<-function(nn){
		tt <- reroot(tree,nn,tree$edge.length[which(tree$edge[,2]==nn)])
		res = fitMk(tt,yy,model=model,fixedQ=Q,output.liks=TRUE)
		#print(paste('tip node:',nn))
		res$lik.anc[1,]
	}
	if(length(nn) > 0) {
		XX<-t(sapply(nn,ff))
		XX<-rbind(YY$lik.anc[1,],XX)
	}
	else {
		XX <- YY$lik.anc[1,,drop=FALSE]
	}
	rownames(XX)<-uncertain
	liks <- YY$lik.anc
	rownames(liks) <- 1:tt$Nnode+n

	return(list(loglik=YY$logLik,Q=Q,marginal.anc=XX,conditional.likelihoods=liks,rerootedTree=tt,tip=uncertain[1]))
}

phylotyper$makeQ <- function(YY) {
	# Compute rate matrix Q from a fitMk object
	Q<-matrix(c(0,YY$rates)[YY$index.matrix+1],length(YY$states),
			length(YY$states),dimnames=list(YY$states,YY$states))
		diag(Q)<--colSums(Q,na.rm=TRUE)

	return(Q)
}

while("phylotyper" %in% search())
  detach("phylotyper")
attach(phylotyper)

File: phylotyper/phylotyper.py
#!/usr/bin/env python

"""Phylotyper module

Phylotyper functions

"""

import logging
import os
import pkg_resources
import rpy2.robjects as robjects
from subprocess import check_output, CalledProcessError, STDOUT


__author__ = "Matthew Whiteside"
__copyright__ = "Copyright 2015, Public Health Agency of Canada"
__license__ = "APL"
__version__ = "2.0"
__maintainer__ = "Matthew Whiteside"
__email__ = "matthew.whiteside@phac-aspc.gc.ca"


class Phylotyper(object):
    """Phylotyper functions.


    """

    def __init__(self, config):
        """Constructor

        Initializes R environment and script settings. Initializes logger.
        
        Args:
            config (obj): PhylotyperConfig with .ini file settings
            
        """

        self.logger = logging.getLogger('phylotyper.phylotyper.Phylotyper')
        self.rwd = pkg_resources.resource_filename(__name__,  '/'.join(('R')))
        self.cwd = os.getcwd()
        self.rlib = config.get('R', 'lib')
        self.rrepo = config.get('R', 'repo')
        self.rscript = config.get('R', 'rscript')
        self.rfiles = {
            'phylotyper': os.path.join(self.rwd, 'phylotyper.R'),
            'tests': os.path.join(self.rwd, 'test_functions.R'),
            'performance_tests': os.path.join(self.rwd, 'performance_tests.R'),
            'subtype_properties': os.path.join(self.rwd, 'subtype_properties.R')
        }


    def subtype(self, tree, subtypes, output_dir, plot_name='posterior_probability_tree.png'):
    	"""Run phylotyper subtyping method

        Wrapper around the phylotyper.R functions. Calls
        R functions using rpy2 interface. Produces png showing 
        posterior probability as pie charts on tree.
        
        Args:
            tree (str): Filepath to newick tree containing both typed & untyped genes
            subtypes (str): Filepath to tab-delimited subtype assignments for typed genes
            output_dir (str): Filepath to output directory
            noplots (bool): Generate png images of tree

        Returns:
            assignments (dict): Each untyped gene key will contain a tuple with:
                1. One or more predicted subtypes
                2. Posterior probability for that assignment
            
        """

        # Set work dir
        robjects.r('setwd("%s")' % self.rwd)

        # Load Phylotyper R functions
        rcode = 'source("%s", chdir=TRUE)' % (self.rfiles['phylotyper']) 
        robjects.r(rcode)

        # Load libraries
        rcode = 'suppressMessages(phylotyper$loadInstallLibraries(libloc="%s",repo="%s"))' % (self.rlib, self.rrepo) 
        robjects.r(rcode)

        # Load data files
        rcode = 'rs = phylotyper$loadSubtype("%s","%s"); tree = rs$tree; subtypes = rs$subtypes; untyped = rs$untyped' % (tree, subtypes)
        robjects.r(rcode)

        untyped = robjects.r('untyped')
        self.logger.debug('phylotyper.R detected the following genes have no subtype: %s' % (','.join(untyped)))

        # Make prior matrix
        rcode = 'priorR = phylotyper$makePriors(tree, subtypes); priorM = priorR$prior.matrix'
        robjects.r(rcode)
       
        # Run subtype procedure
        rcode = 'est.scheme = 5; result = phylotyper$runSubtypeProcedure(tree, priorM, est.scheme, tips=untyped)'
        robjects.r(rcode)

        # Write subtype predictions
        rcode = '''
        cn = colnames(result$tip.pp);
        matrix(round(result$tip.pp[untyped,], digits=7),ncol=length(cn),nrow=length(untyped),dimnames=list(untyped,cn), byrow=TRUE)
        '''
        predictions = robjects.r(rcode)
        subtype_states = predictions.colnames
        assignments = {}

        # Find largest pp
        for genome in untyped:
            pprow = predictions.rx(genome, True)
            
            maxpp = -1
            maxst = []
            i = 0
            for pp in pprow:
                if pp > maxpp:
                    maxst = [subtype_states[i]]
                    maxpp = pp
                elif pp == maxpp:
                    maxst.append(subtype_states[i])

                i += 1

            assignments[genome] = (maxst, maxpp)

        # Plot pp on the tree
        if plot_name:
            rcode = '''
            dim = phylotyper$plotDim(tree)
            graphics.off()
            png(filename="%s", width=dim[['x']],height=dim[['y']],res=dim[['res']])
            do.call(result$plot.function, list(tree=tree, fit=result$result, subtypes=subtypes))
            graphics.off()
            ''' % (plot_name)
            robjects.r(rcode)

        # Return assignments
        robjects.r('setwd("%s")' % self.cwd)
        return assignments
  		

    def run(self, tree, subtypes, output_dir):
        """Run phylotyper R script
        
        Calls external R script
        
        Args:
            tree (str): Filepath to newick tree containing both typed & untyped genes
            subtypes (str): Filepath to tab-delimited subtype assignments for typed genes
            output_dir (str): Filepath to output directory

        Returns:
            results are output to output_dir
            
        """

        cmd = "{} {} {} {} --out {}".format(self.rscript, self.rfiles['performance_tests'],
            tree, subtypes, output_dir)
        cwd = os.getcwd()
        os.chdir(self.rwd)

        try:
            check_output(cmd, stderr=STDOUT, shell=True, universal_newlines=True)                         
        except CalledProcessError as e:
            os.chdir(cwd)
            msg = "R script {} failed: {} (return code: {}).".format(cmd, e.output, e.returncode)                                                                                                   
            raise Exception(msg)

        os.chdir(cwd)
        return None


    def evaluate(self, tree, subtypes, output_dir):
        """Evaluate new subtype scheme
        
        Runs performance tests and reports accuracy. Identifies anomylous subtypes.
        Builds transition matrix to use in subtyping.
        
        Args:
            tree (str): Filepath to newick tree containing both typed & untyped genes
            subtypes (str): Filepath to tab-delimited subtype assignments for typed genes
            output_dir (str): Filepath to output directory

        Returns:
            results are output to output_dir
            
        """

        # Set work dir
        robjects.r('setwd("{}")'.format(self.rwd))

        # Load Phylotyper R functions
        rcode = 'source("{}")'.format(self.rfiles['phylotyper']) 
        robjects.r(rcode)

        # Load libraries
        rcode = 'suppressMessages(phylotyper$loadInstallLibraries(libloc="{}",repo="{}"))'.format(self.rlib, self.rrepo) 
        robjects.r(rcode)

        # Load test functions
        rcode = 'source("{}"); source("{}");'.format(self.rfiles['tests'], self.rfiles['subtype_properties']) 
        robjects.r(rcode)

        # Load data files
        rcode = 'rs = phylotyper$loadSubtype("{}","{}"); tree = rs$tree; subtypes = rs$subtypes'.format(tree, subtypes)
        robjects.r(rcode)

        # Check for weird subtypes
        rcode = '''
        fixes = reassign.subtypes(tree, subtypes);
        row.names(fixes)[as.character(fixes$new) != as.character(fixes$prev)]
        '''
        suspect = robjects.r(rcode)

        if suspect:
            newvalues = robjects.r('as.character(fixes$new)')
            genomes = robjects.r('row.names(fixes)')
            accession_map = 
           
            # Save to file
            newsubtype_file = os.path.join(output_dir, 'phylotyper_proposed_subtypes.csv')
            with open(newsubtype_file, 'w') as outfh:
                for i in xrange(len(newvalues)):
                    outfh.write('{}\t{}\n'.format(genomes[i], newvalues[i]))

            self.logger.warn(
                '''\nSuspicious Subtypes! Detected anomylous subtype assignments for the following sequences
                that do not follow phylogenetic groupings:
                  {}
                A proposed subtype assignment has been generated and is available in:
                  {}
                Review the subtypes and if you want to use this subtype assignment, replace your subtype file 
                with this one and re-run the build script.
                '''.format('\n  '.join(tuple(suspect)), newsubtype_file, newsubtype_file))



        # untyped = robjects.r('untyped')
        # self.logger.debug('phylotyper.R detected the following genes have no subtype: %s' % (','.join(untyped)))

        # # Make prior matrix
        # rcode = 'priorR = phylotyper$makePriors(tree, subtypes); priorM = priorR$prior.matrix'
        # robjects.r(rcode)
       
        # # Run subtype procedure
        # rcode = 'est.scheme = 5; result = phylotyper$runSubtypeProcedure(tree, priorM, est.scheme, tips=untyped)'
        # robjects.r(rcode)

        # # Write subtype predictions
        # rcode = '''
        # cn = colnames(result$tip.pp);
        # matrix(round(result$tip.pp[untyped,], digits=7),ncol=length(cn),nrow=length(untyped),dimnames=list(untyped,cn), byrow=TRUE)
        # '''
        # predictions = robjects.r(rcode)
        # subtype_states = predictions.colnames
        # assignments = {}

        # # Find largest pp
        # for genome in untyped:
        #     pprow = predictions.rx(genome, True)
            
        #     maxpp = -1
        #     maxst = []
        #     i = 0
        #     for pp in pprow:
        #         if pp > maxpp:
        #             maxst = [subtype_states[i]]
        #             maxpp = pp
        #         elif pp == maxpp:
        #             maxst.append(subtype_states[i])

        #         i += 1

        #     assignments[genome] = (maxst, maxpp)

        # # Plot pp on the tree
        # if plot_name:
        #     rcode = '''
        #     dim = phylotyper$plotDim(tree)
        #     graphics.off()
        #     png(filename="%s", width=dim[['x']],height=dim[['y']],res=dim[['res']])
        #     do.call(result$plot.function, list(tree=tree, fit=result$result, subtypes=subtypes))
        #     graphics.off()
        #     ''' % (plot_name)
        #     robjects.r(rcode)

        # # Return assignments
        # return assignments

        robjects.r('setwd("%s")' % self.cwd)




        


       


  
	